kong/pdk/node.lua

```diff
@@ -258,20 +258,33 @@ local function new(self)
   end
 
 
+  -- the PDK can be even when there is no configuration (for docs/tests)
+  -- so execute below block only when running under correct context
   local prefix = self and self.configuration and self.configuration.prefix
-  if prefix and self.configuration.role == "data_plane" then
-    local id, err = private_node.load_node_id(prefix)
-    if id then
-      node_id = id
-      ngx.log(ngx.DEBUG, "restored node_id from the filesystem: ", node_id)
-
-    else
-      id = _NODE.get_id()
-      if err then
-        ngx.log(ngx.WARN, "failed to restore node_id from the filesystem: ",
-                err, ", so a new one was generated: ", id)
+  local configuration_node_id = self and self.configuration and self.configuration.node_id
+  if prefix  then
+    -- precedence order:
+    -- 1. user provided node id
+    if configuration_node_id then
+      node_id = configuration_node_id
+    end
+    -- 2. node id (if any) on file-system
+    if not node_id then
+      if prefix and self.configuration.role == "data_plane" then
+        local id, err = private_node.load_node_id(prefix)
+        if id then
+          node_id = id
+          ngx.log(ngx.DEBUG, "restored node_id from the filesystem: ", node_id)
+        else
+          ngx.log(ngx.WARN, "failed to restore node_id from the filesystem: ",
+                  err, ", a new node_id will be generated")
+        end
       end
     end
+    -- 3. generate a new id
+    if not node_id then
+      _NODE.get_id()
```

Comment:
i think this can be preserved in the `else` branch above, so the new id is logged, as it was

Reply:
I did this to improve the readability of the code a bit. Is logging the new ID necessary?
If so, I can add another log at the end of these conditional blocks to log it unconditionally.

Reply:
IMO it's questionable that it achieves the goal of improving readability -- it introduces an unnecessary conditional. But let's not bikeshed over this; I can approve / merge if you are adamant on the change.

Reply:
Let's keep it as is and move on.

Reply:
I added a log at INFO level for node-id.